specifyweb/frontend/js_src/lib/components/TreeView/AddRank.tsx

```diff
@@ -28,128 +27,79 @@ export function AddRank<SCHEMA extends AnyTree>({
-  const [isAddingParentRank, setIsAddingParentRank] = React.useState(false);
+  const [state, setState] = React.useState<'add' | 'initial' | 'parent'>(
+    'initial'
+  );
 
   const [parentRank, setParentRank] = React.useState('');
-  const treeId = strictIdFromUrl(treeDefinitionItems[0].treeDef);
 
   const treeResource = React.useMemo(
-    () =>
-      tableName === 'Geography'
-        ? new tables.GeographyTreeDefItem.Resource({
-            _tableName: 'GeographyTreeDefItem',
-          })
-        : tableName === 'Taxon'
-        ? new tables.TaxonTreeDefItem.Resource({
-            _tableName: 'TaxonTreeDefItem',
-          })
-        : tableName === 'LithoStrat'
-        ? new tables.LithoStratTreeDefItem.Resource({
-            _tableName: 'LithoStratTreeDefItem',
-          })
-        : new tables.StorageTreeDefItem.Resource({
-            _tableName: 'StorageTreeDefItem',
-          }),
+    () => new tables[treeDefinitionItems[0]._tableName].Resource(),
     [tableName]
   );
 
-  function addRank(): void {
-    const url = `/api/specify_tree/${tableName.toLowerCase()}/add_tree_rank/`;
-    loading(
-      ping(url, {
-        method: 'POST',
-        body: {
-          newRankName: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('name'),
-          parentRankName: parentRank,
-          treeID: treeId,
-          newRankTitle: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('title'),
-          remarks: (treeResource as SpecifyResource<GeographyTreeDefItem>).get(
-            'remarks'
-          ),
-          textAfter: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('textAfter'),
-          textBefore: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('textBefore'),
-          isEnforced: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('isEnforced'),
-          isInFullName: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('isInFullName'),
-          fullNameSeparator: (
-            treeResource as SpecifyResource<GeographyTreeDefItem>
-          ).get('fullNameSeparator'),
-        },
-      }).then(() => globalThis.location.reload())
-    );
-    setIsAddingRank(false);
-  }
-
   return (
     <>
       <Button.Icon
         icon="plus"
         title={treeText.addNewRank()}
-        onClick={() => setIsAddingParentRank(true)}
+        onClick={() => setState('parent')}
       />
-      {isAddingParentRank && (
+      {state === 'parent' && (
         <Dialog
           buttons={
             <>
               <Button.DialogClose>{commonText.cancel()}</Button.DialogClose>
               <Button.Save
                 onClick={() => {
-                  setIsAddingRank(true);
-                  setIsAddingParentRank(false);
+                  const resourceParent = React.useMemo(
```

Comment:
remember rules of hooks 🚨🚨🚨

https://legacy.reactjs.org/docs/hooks-rules.html

Reply:
what's the point of this useMemo in here at all?

As a reminder, `useMemo` is to prevent re-execution of some expression during re-renders of the same component.
For react to know clearly what was the value of that hook during the previous render, all hooks must be called in the same order, and not inside a condition/callback/async